[Blazor] Add `IRazorComponentsBuilder.AddWebAssemblyComponents()`

MackinnonBuck opened the pull request:
## Add `IRazorComponentsBuilder.AddWebAssemblyComponents()`

Adds the `IRazorComponentsBuilder.AddWebAssemblyComponents()` extension method. This allows Blazor WebAssembly components to be rendered from a Blazor app that utilizes SSR.

Fixes #46354

MackinnonBuck committed:
Add 'AddWebAssemblyComponents()'

SteveSandersonMS commented on src/Components/WebAssembly/Server/src/Microsoft.AspNetCore.Components.WebAssembly.Server.csproj:
```diff
@@ -12,6 +12,7 @@
   </PropertyGroup>
 
   <ItemGroup>
+    <Reference Include="Microsoft.AspNetCore.Components.Endpoints" />
```
It might not make any difference, and there's probably an important reason you did it this way, but TBH I was expecting the dependency be Endpoints->WebAssembly.Server rather than this way around.

That's basically because WebAssembly.Server is an existing, self-contained thing which doesn't need anything from Endpoints to do its existing job, whereas Endpoints is a kind of aggregation of our functionality from many areas so it would be intuitive for it to pull in WebAssembly.Server to achieve the new functionality.

If it is possible to swap this around and have `AddWebAssemblyComponents` live inside `Endpoints`, that would seem more obvious. But if that's impractical or undesirable for some reason I'd be interested to understand!

MackinnonBuck replied:
Having the dependency be Endpoints->WebAssembly.Server means we're pulling WebAssembly.Server into the shared framework. I wasn't sure if there was a specific reason why WebAssembly.Server wasn't already in the shared framework, but I don't see a reason why it shouldn't be included.

I've now swapped the direction of this dependency and added WebAssembly.Server into the shared framework, but I'm happy to revert it back if there's a reason we don't want to make that change 🙂

SteveSandersonMS replied:
> added WebAssembly.Server into the shared framework

That sounds good to me. I don't think there was any particular benefit in making people reference that package explicitly in the past.

SteveSandersonMS replied:
We discussed this separately and found we couldn't really invert the dependency here because that would pull other transitive dependencies into the shared framework. Marking this discussion as resolved. We may revisit this in the future.

SteveSandersonMS reviewed (approved):
Approving because it's clear this is close to being ready and you can decide for yourself whether it's right to swap around the dependency direction as per my comment.

Also I think we're now at a point where the E2E tests for Endpoint rendering could be extended to include some cases where we insert interactive Server/WebAssembly components and observe that they are actually interactive. Since we already have the core structure for those E2E tests (i.e., there's an app, and the tests know how to start it, etc.) it might not be a massive amount of extra work to introduce one or two tests to show interactivity is possible now. Would you consider that? If you feel it's outside the scope of what you want to handle here that's fine - we just have to get to that at some point :)

MackinnonBuck committed:
PR feedback + new APIs + basic E2E tests

MackinnonBuck commented on src/Components/Endpoints/src/DependencyInjection/WebAssemblyComponentsEndpointOptions.cs:
```diff
@@ -0,0 +1,17 @@
+
+namespace Microsoft.AspNetCore.Components.Endpoints;
+
+/// <summary>
+/// Options to configure interactive WebAssembly components in a razor components application.
+/// </summary>
+public sealed class WebAssemblyComponentsEndpointOptions
+{
+    /// <summary>
+    /// Gets or sets <see cref="PathString"/> that indicates the prefix for Blazor WebAssembly framework files.
+    /// </summary>
+    public PathString FrameworkFilesPathPrefix { get; set; }
+}
```
This API allows the path prefix passed to `UseBlazorFrameworkFiles` to be configurable in a "full-stack Blazor"-style app.

javiercn replied:
Is this something we could avoid?

MackinnonBuck replied:
What's the reason for wanting to avoid it? Is it to keep the API smaller for the initial release? Without this API, I don't think there's a way for customers to configure the path prefix passed to `UseBlazorFrameworkFiles()`. I was using this API in our own test app so that `Components.TestServer` could reference both `BasicTestApp` as well as a new, minimal WebAssembly project.

SteveSandersonMS replied:
Seems reasonable to allow this configurability if it's what's need to host multiple independent WebAssembly apps on a single server app.

MackinnonBuck replied:
This only impacts the path from which `_framework` folder content is loaded. The paths from which `wwwroot` content gets served can be configured with `UseStaticFiles()`.

I'm worried `PathPrefix` might not be clear enough in the context of calling `AddWebAssemblyComponents()`. I think `StaticAssetsPathPrefix` is a bit too general given that it's specifically `_framework` files that we're configuring.

Although, I can't think of a case where the `_framework` path prefix would be different than the path prefix for `wwwroot` content. So maybe it's okay to go with `StaticAssetsPathPrefix`? But maybe then there could still be some confusion around whether calling `UseStaticFiles()` is still necessary.

Perhaps we keep the current name but update the XML comment to make this clearer?

SteveSandersonMS replied:
It also affects `_content`, doesn't it? Or if not I think that would be at risk of clashing. And possibly any other framework-supplied files such as hot reload scripts, though I'm not so sure about that.

I don't feel strongly about the name, so if you think `FrameworkFilesPathPrefix` is the safest choice in terms of accuracy and understandability, that's fine with me.

MackinnonBuck replied:
Let's do `PathPrefix` for now - one could make the argument that the fact that it only affects `_framework` files is an uninteresting detail to the developer.

SteveSandersonMS replied:
> I could totally be missing something, but from what I can tell it looks like _content isn't affected by this:

Is that a bug then? I thought the scenario here was "hosting unrelated WebAssembly apps on separate paths". If that's the case we can't have them both using `/_content`.

MackinnonBuck commented on src/Components/test/testassets/Components.TestServer/RazorComponentEndpointsStartup.cs:
```diff
@@ -22,7 +22,12 @@ public RazorComponentEndpointsStartup(IConfiguration configuration)
-        services.AddRazorComponents();
+        services.AddRazorComponents()
+            .AddServerComponents()
+            .AddWebAssemblyComponents(options =>
+            {
+                options.FrameworkFilesPathPrefix = "/WasmMinimal";
+            });
```
This is so that we don't inherit the baggage from `BasicTestApp` when all we need is a super minimal WebAssembly app to enable WebAssembly interactivity.

MackinnonBuck commented on src/Components/test/testassets/Components.WasmMinimal/Program.cs:
```diff
@@ -0,0 +1,10 @@
+using System.Reflection;
+using Microsoft.AspNetCore.Components.WebAssembly.Hosting;
+
+Assembly.Load(nameof(TestContentPackage));
```
Without this, the `TestContentPackage` assembly does not get loaded, and [`RootComponentTypeCache`](https://github.com/dotnet/aspnetcore/blob/6575b289081b9f85fb30d01c7b306d1ec5d00a68/src/Components/Shared/src/RootComponentTypeCache.cs#LL24C22-L24C22) can't discover the component.

Would we consider updating `RootComponentTypeCache` to attempt to load an assembly if it couldn't be found?

SteveSandersonMS replied:
This is very interesting. Is this just a quirk of the framework that is already the case in .NET 7? For example, if you used `<component ... render-mode="WebAssembly">` (as opposed to declaring the root component in `Program.cs`) could you get the same issue? My guess is yes.

Rather than block your PR on it, I would recommend filing an issue for us to dig into (unless you think you are actively *causing* this issue in your PR here, which doesn't appear to be the case to me).

> Would we consider updating RootComponentTypeCache to attempt to load an assembly if it couldn't be found?

Assuming you mean code that is actually running on WebAssembly (e.g., inside `WebAssemblyHostBuilder`) then yes sure, why not. It's not a security concern in that case.

There could still be a problem since trimming may have removed the component if there was no statically-observable evidence that it was going to be used. However in practice that would be uncommon since we don't trim user assemblies by default. This may change in the future.

MackinnonBuck commented on src/Components/test/testassets/Components.TestServer/RazorComponents/Components/WebAssemblyInteractiveCounter.razor:
This component isn't actually used in the E2E tests because this project isn't multi-targeted. I put it here anyway so that there's a WebAssembly component to get picked up by [component discovery](https://github.com/dotnet/aspnetcore/blob/6575b289081b9f85fb30d01c7b306d1ec5d00a68/src/Components/Endpoints/src/Discovery/RazorComponentApplication.cs#L37), which enables the `WebAssemblyEndpointProvider`. It's a bit of a hack but I'm not aware of a better workaround at the moment.

@javiercn It doesn't appear that [`RazorComponentDataSourceOptions`](https://github.com/dotnet/aspnetcore/blob/6575b289081b9f85fb30d01c7b306d1ec5d00a68/src/Components/Endpoints/src/Builder/RazorComponentDataSourceOptions.cs) can be configured by application code yet. Do we intend on adding APIs for that?

MackinnonBuck commented:
Thanks for the feedback, @SteveSandersonMS! I've added some basic E2E tests to validate interactivity.

I'm re-requesting a review because I've added some new APIs since you last looked (and brought `WebAssembly.Server` into the shared framework).

javiercn commented on src/Components/Endpoints/src/DependencyInjection/WebAssemblyRazorComponentsBuilderExtensions.cs:
```diff
@@ -0,0 +1,113 @@
+    /// <param name="builder">The <see cref="IRazorComponentsBuilder"/>.</param>
+    /// <param name="configure">A callback to configure <see cref="WebAssemblyComponentsEndpointOptions"/>.</param>
+    /// <returns>An <see cref="IRazorComponentsBuilder"/> that can be used to further customize the configuration.</returns>
+    public static IRazorComponentsBuilder AddWebAssemblyComponents(this IRazorComponentsBuilder builder, Action<WebAssemblyComponentsEndpointOptions>? configure = null)
```
Why does this not go on `Webassembly.Server`?

MackinnonBuck replied:
😄 https://github.com/dotnet/aspnetcore/pull/48664#discussion_r1222680088

MackinnonBuck replied:
I initially put it in `WebAssembly.Server` because it's consistent with how `AddServerComponents` is in `Components.Server`, and inverting the direction of the dependency requires moving `WebAssembly.Server` into the shared framework.

Does this align with your thinking, or was there another reason you think it belongs in `WebAssembly.Server`, @javiercn?

MackinnonBuck committed:
Update Microsoft.AspNetCore.Components.WebAssembly.Server.csproj

amcasey commented:
Hard to see how the CI failure in `Http2WebSocketTests.HEADERS_Received_SecondRequest_ConnectProtocolReset()` could be related - probably flaky.

MackinnonBuck committed:
Keep shipping WebAssembly.Server as a package

MackinnonBuck committed:
Fix tests

MackinnonBuck committed:
Update Framework.slnf

MackinnonBuck commented:
I can't figure out why `Microsoft.AspNetCore.Components.WebAssembly.Server` is not getting picked up during source build.

@wtgodbe Do you know why this could be happening?

wtgodbe commented:
> I can't figure out why `Microsoft.AspNetCore.Components.WebAssembly.Server` is not getting picked up during source build.
> 
> @wtgodbe Do you know why this could be happening?

Some of our custom reference resolution logic is turned off for source build, so it may be referencing an older version of the project. The source build binlog will probably contain useful info about what's getting resolved here. @MichaelSimons is this the kind of situation we've used source build patches for in the past?

MichaelSimons commented:
> Some of our custom reference resolution logic is turned off for source build, so it may be referencing an older version of the project. The source build binlog will probably contain useful info about what's getting resolved here. 

Taking a cursory look, the WebAssembly components where it looks like UseBlazorFrameworkFiles is defined are excluded from SourceBuild - https://github.com/dotnet/aspnetcore/blob/main/src/Components/WebAssembly/Directory.Build.props.

> @MichaelSimons is this the kind of situation we've used source build patches for in the past?

No

MackinnonBuck committed:
Create Directory.Build.props

MackinnonBuck committed:
Move AddWebAssemblyComponents back to WebAssembly.Server

MackinnonBuck committed:
Update SharedFramework.Local.props

MackinnonBuck committed:
Update PublicAPI.Unshipped.txt

MackinnonBuck commented:
I've moved `AddWebAssemblyComponents` back to `WebAssembly.Server` because there were some challenges with bringing `WebAssembly.Server` into the shared framework at this time.

@SteveSandersonMS @javiercn Do you think this is good to merge?

javiercn commented on src/Components/WebAssembly/Server/src/RazorComponentsBuilderExtensions.cs:
```diff
@@ -0,0 +1,113 @@
+        return builder;
+    }
+
+    private class WebAssemblyEndpointProvider : RenderModeEndpointProvider
```
We are missing two extension methods `'AddWebAssemblyRenderMode', 'AddServerRenderMode'` to plug a render mode independent of discovery.

MackinnonBuck replied:
Do you think it's fine if I add those as part of this PR?

MackinnonBuck replied:
Looks like those are [already a thing](https://github.com/dotnet/aspnetcore/blob/3f389bb07e872cd11b7b7b401ec673465bf89600/src/Components/Endpoints/src/Builder/RazorComponentEndpointConventionBuilder.cs#L47), actually.

MackinnonBuck committed:
Use Add{Server|WebAssembly}RenderMode()

SteveSandersonMS commented on src/Components/Endpoints/src/DependencyInjection/WebAssemblyComponentsEndpointOptions.cs:
```diff
@@ -0,0 +1,17 @@
+namespace Microsoft.AspNetCore.Components.Endpoints;
+
+/// <summary>
+/// Options to configure interactive WebAssembly components in a razor components application.
```
```suggestion
/// Options to configure interactive WebAssembly components.
```

Just a nit but the terminology gets difficult otherwise. We have so many different names (components, Razor components, Blazor) before getting into the differences between the older and newer hosting styles.

I think we're pushing towards not differentiating types of ASP.NET Core applications and so don't have to caveat what type of app the developer needs to be working with.

SteveSandersonMS reviewed (approved):
Looks great!

MackinnonBuck committed:
Update src/Components/Endpoints/src/DependencyInjection/WebAssemblyComponentsEndpointOptions.cs

Co-authored-by: Steve Sanderson <SteveSandersonMS@users.noreply.github.com>

MackinnonBuck committed:
Rename FrameworkFilesPathPrefix to PathPrefix
